pkg/fanal/image/daemon/image.go

```diff
@@ -243,6 +243,12 @@ func emptyLayer(history dimage.HistoryResponseItem) bool {
 		strings.HasPrefix(createdBy, "ARG") {
 		return true
 	}
+	// For "echo" statements in the image creation (e.g. RUN echo "hello world"), docker treats the layer as empty,
+	// but podman and BUILDKIT mark it as non empty layer
+	// So this condition in the below conditional statement handles it for images built with docker
+	if strings.HasPrefix(strings.ToUpper(createdBy), "ECHO") {
```

Comment:
Do you mean the instruction will be `ECHO` rather than `RUN` in Docker? Podman and Buildkit put the `RUN` instruction though.

Reply:
The created by in docker is:
/bin/sh -c echo "hello world"

![image](https://user-images.githubusercontent.com/33202950/204497960-0ea02fbd-41c1-48d4-ad6d-2d023912df9c.png)

For buildkit it is:
RUN /bin/sh -c echo "hello world" # buildkit

![image](https://user-images.githubusercontent.com/33202950/204498186-87521eed-3b52-4c9e-8de8-c3d563225739.png)

Reply:
I think the logic is problematic.

```
    {
      "created": "2022-12-11T13:42:06.570705125Z",
      "created_by": "/bin/sh -c echo 'hello' >> hello.txt"
    },
    {
      "created": "2022-12-11T13:42:07.804305782Z",
      "created_by": "/bin/sh -c echo 'hello'",
      "empty_layer": true
    },
```

We should look into the Docker source code.

Reply:
We should see if this logic helps us.
https://github.com/aquasecurity/trivy/blob/e8631ea7908812239a0911173ee15798e65eff5a/pkg/fanal/image/daemon/image.go#L227-L229

Reply:
We added this change because for local images built with Docker; the step 'RUN echo "hello world"' size is 0.
So the below condition will be false 
 if history.Size != 0 { 
 	return false 
 } 
 
 But when it goes to the check; the given conditions do not cover 'RUN echo "hello world"' - the layer is returned as non-empty. 
https://github.com/aquasecurity/trivy/blob/4436a202ff2f4f8edc12a62cc0a1b0a9905a70e7/pkg/fanal/image/daemon/image.go#L230-L245


The scenario that you mentioned:
    {
      "created": "2022-12-11T13:42:06.570705125Z",
      "created_by": "/bin/sh -c echo 'hello' >> hello.txt"
    },
    {
      "created": "2022-12-11T13:42:07.804305782Z",
      "created_by": "/bin/sh -c echo 'hello'",
      "empty_layer": true
    },

for the first case - "created_by": "/bin/sh -c echo 'hello' >> hello.txt";  "if history.Size != 0 {  " condition will handle it.
For the second case : "created_by": "/bin/sh -c echo 'hello'"; I added the line in this PR. 

Please let me know if I could explain myself

Reply:
It is ok then. Thanks for confirming.

Reply:
How about Podman? Does this logic handle the `echo` layer created by Podman properly?

Reply:
For podman too, echo is a non-empty layer
![image](https://user-images.githubusercontent.com/33202950/207397545-198bc606-0581-438c-bf83-732b5df9f590.png)

Reply:
Isn't it a problem? The current logic returns true when the created by is `"/bin/sh -c echo \"hello world\"",`. But we should return false for Podman. Am I missing something?

Reply:
For Podman, similar to BUILDKIT, "if history.Size != 0 { " this will handle..

Reply:
Not sure, I saw the same behaviour with both Podman and Buildkit . I have used exactly the same steps to build all the images

Reply:
It makes sense the size could be more than zero with `echo hello > hello.txt`, but I'm curious why it will not be zero with `echo hello`.
Anyway, you confirmed it will not be zero in Podman and we can handle it by `if history.Size != 0`, right?

Reply:
Yes. I had tested for the same.
But today we found another instance of similar issue

![image](https://user-images.githubusercontent.com/33202950/207887444-694d237f-7e90-4b06-8320-f199122d2f58.png)

I think we might keep this PR on hold to have a better fix post analysis on similar cases?

Reply:
Seems like COPY should be always non-empty, but do you mean it is empty?
https://github.com/moby/buildkit/blob/2942d13ff489a2a49082c99e6104517e357e53ad/frontend/dockerfile/dockerfile2llb/convert.go#L1221